### PR TITLE
Add error_on_line_overflow_strings config option

### LINKED ISSUE #2268: A URL (or any other text containing `//`) incorrectly sets `is_comment` to true
## What happened 

I have a project containing this constant:

```rust
const RFC_CRATE_FILE_EXAMPLE: &[u8] = br#"{"name":"my_serde","vers":"1.0.11","deps":[{"name":"serde","req":"^1.0","registry":"https://github.com/rust-lang/crates.io-index","features":[],"optional":true,"default_features":true,"target":null,"kind":"normal"}],"cksum":"f7726f29ddf9731b17ff113c461e362c381d9d69433f79de4f3dd572488823e9","features":{"default":["std"],"derive":["serde_derive"],"std":[]},"yanked":false}
"#;
```

Using rustfmt-nightly 0.2.7 and Rust nightly 2017-11-07, if I run:

```
cargo +nightly-2017-11-07 fmt
```

I get this output from rustfmt:

```
error: line exceeded maximum width (maximum: 100, found: 414)
 --> /Users/carolnichols/rust/rustfmt-test/src/lib.rs:1
  |
1 | const RFC_CRATE_FILE_EXAMPLE: &[u8] = br#"{"name":"my_serde","vers":"1.0.11","deps":[{"name":"serde","req":"^1.0","registry":"https://github.com/rust-lang/crates.io-index","features":[],"optional":true,"default_features":true,"target":null,"kind":"normal"}],"cksum":"f7726f29ddf9731b17ff113c461e362c381d9d69433f79de4f3dd572488823e9","features":{"default":["std"],"derive":["serde_derive"],"std":[]},"yanked":false}
  |                                                                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: use `error_on_line_overflow_comments = false` to suppress the warning against line comments

warning: rustfmt may have failed to format. See previous 1 errors.
```

It says to suppress this warning, I should use `error_on_line_overflow_comments = false`, but this line isn't a comment.

## What I expected

I expected rustfmt to tell me this line is too long, but instead of telling me to use `error_on_line_overflow_comments = false`, I expected it to tell me to use `error_on_line_overflow = false` since this line isn't a comment.

## Likely cause

I believe this is because of the `//` that appears in the URL in my raw byte string, and the line length code isn't actually parsing the Rust code, so it [sets `is_comment` to true](https://github.com/rust-lang-nursery/rustfmt/blob/68b43b2240fe03edab4d0951a1d4c9bb04b559f4/src/lib.rs#L445-L447) if it sees two `/` characters in a row anywhere on a line.

### PR DESCRIPTION
**Pull request by topecongiro**
This PR adds a new config option `error_on_line_overflow_strings`, which is a variant of `error_on_line_overflow_comments`.

EDIT: closes #2268.

### CODE REVIEW

**Comment by topecongiro** [comment]
I have a few questions:

1. `error_on_line_overflow_strings` and `error_on_line_overflow_strings` suppress warnings on trailing white spaces as well. To make things clear, should we change the name of two options?
2. Should we combine the two options?
3. Should we turn them off (set to `false`) by default?

File: src/config.rs [hunk 1]
```diff
@@ -674,6 +674,8 @@ create_config! {
     error_on_line_overflow: bool, true, false, "Error if unable to get all lines within max_width";
     error_on_line_overflow_comments: bool, true, false,
         "Error if unable to get comments within max_width";
+    error_on_line_overflow_strings: bool, true, false,
+        "Error if unable to get string letrais within max_width";
```
Comment (CAD97) [inline src/config.rs]: `letrais` does not match any dictionary word.

Did you mean one of:

- `letters`
- `literal`
- `laterals`
- `literals`
- `lateral`

?

(Sorry, just having fun. Those are legitimate suggestions from a spell checker, though. You meant `literals` here.)

Comment (topecongiro) [reply src/config.rs]: Thank you, this will make my typo of the year for 2017.

**Comment by nrc** [comment]
>    error_on_line_overflow_strings and error_on_line_overflow_strings suppress warnings on trailing white spaces as well. To make things clear, should we change the name of two options?

Yeah, I think `error_on_unformatted` would be good, to cover everything.

>   Should we combine the two options?

Yes, I can't imagine someone wanting one, but not the other

>   Should we turn them off (set to false) by default?

Hmm, not sure about the default. I think probably yes the option should be off by default. But we should be able to turn it on via the command line (for the CI use case).

**Comment by topecongiro** [comment]
Updated the PR with the review reflected and rebased against master.

1. Combined the two options into `error_on_unformatted_comments_or_strings` (too long?)
2. Added `--error-on-unformatted` command line option.
3. Sorted command line options in alphabetical order.

```


usage: ./target/debug/rustfmt [options] <file>...

Options:
        --color [always|never|auto]
                        Use colored output (if supported)
        --config-help   Show details of rustfmt configuration options
        --config-path [Path for the configuration file]
                        Recursively searches the given path for the
                        rustfmt.toml config file. If not found reverts to the
                        input file path
        --dump-default-config PATH
                        Dumps default configuration to PATH. PATH defaults to
                        stdout, if omitted.
        --dump-minimal-config PATH
                        Dumps configuration options that were checked during
                        formatting to a file.
        --error-on-unformatted 
                        Error if unable to get comments or string literals
                        within max_width, or they are left with trailing
                        whitespaces
        --file-lines JSON
                        Format specified line ranges. See README for more
                        detail on the JSON format.
    -h, --help          Show this message
        --skip-children 
                        Don't reformat child modules
        --unstable-features 
                        Enables unstable features. Only available on nightly
                        channel
    -v, --verbose       Print verbose output
    -V, --version       Show version information
        --write-mode [replace|overwrite|display|plain|diff|coverage|checkstyle]
                        How to write output (not usable when piping from
                        stdin)

Exit Codes:
    0 = No errors
    1 = Encountered operational errors e.g. an IO error
    2 = Failed to reformat code because of parsing errors
    3 = Code is valid, but it is impossible to format it properly
    4 = Formatted code differs from existing code (write-mode diff only)
```

**Comment by nrc** [comment]
> Combined the two options into error_on_unformatted_comments_or_strings (too long?)

Could we use just `error_on_unformatted` for the option too? I think it would be good to be consistent between the option and command line flag

**Comment by nrc** [comment]
r+ with that last change

**Comment by topecongiro** [comment]
Thanks again, rename the option.

**Comment by topecongiro** [comment]
Rebased against master.

**Comment by nrc** [comment]
Thanks!

For the future, it would be great if you could squash some of this kind of commit - exactly how to structure commits in a PR is more art than science (and yours are usually very nice, and I am not perfect by a long way). I would squash typo and minor refactoring commits into commits making larger changes, though I would try and keep refactoring commits separate from those that change behaviour.  The best guide, I think, is to consider if when reading blame for the repo, whether it is useful to have the code in a separate commit - to take some examples from this PR - 'Print command line options in alphabetical order' and 'Add a config option to suppress error message on string literal' - definitely yes, 'Fix a typo ' - definitely no, 'Organize command line options and start with upper case' - maybe or maybe better squashed with 'Print command line options in alphabetical order', depending on the code. Anyway, this doesn't have to be perfect, and I wouldn't spend too much time on it, but there might be a few places you can make a better commit series without too much stress.

**Comment by topecongiro** [comment]
Thank you the additional advice, this is very helpful!

**Pull request merged**